프로필 관련 뷰 구현

KWSMooBang opened the pull request:
### 프로필 표시 관련 뷰
- [x] Multi Section TableView구현을 위한 Section 및 Item 모델 구현 (DataSourceModel파일)
- [x] ProfileTabView, EditProfileView, DetailProfileView의 TableView binding 및 header,footer 구현 

### 프로필 수정 관련 뷰
- [ ] AddInformationView 동작 구현
- [x] SelectInformationView 동작 구현
- [x] 프로필 수정 관련 TableViewCell 구현

Commit:
EditProfileView Cell bind

EditProfileView의 tableview에 cell bind를 위한 SectionModel 및 Item 생성, header, cell, footer bind

ProfileTab TableView Cell button, tap bind fix

Profile Tab View Tableview bind

Add EditDetailInformationView and Controller

Add AddInformationView, SelectInformationView

same commit

Add DateSelectTableViewCell

SelectInformationView pass data

change some cells

commit about DateSelectCell

add DataSelectSectionHeaderView

test commit

프로필 관련 Cell 추가 구현

자기 소개 편집 뷰 추가

cell configure 단순화

cell 원형 이미지 수정

대학, 직장 추가 뷰 관련 기능 추가

대학,직장 추가 뷰 사이 데이터 전달 구현

프로필 편집 관련 뷰, 생일 성별 선택 뷰 추가

프로필 및 이름 편집 뷰 구현

프로필 탭 유저 프로필 데이터 불러오기 구현

프로필 탭 유저 프로필 데이터 불러오기

프로필 관련 뷰 데이터 로드 구현

자기소개 추가 시 alert형식의 메뉴 구현

생일 및 성별 수정 기능 구현

자기소개 편집 구현

이름 변경 기능 구현

프로필 탭 자기소개 표시

페이지 네이션 모델 로그인 삭제

직장, 대학 정보 추가 뷰 기능 일부 구현

프로필 관련 네트워크 수정

프로필 및 커버 사진 업로드

아이콘 이미지 변경

파일 이름 에러 수정

시작 화면 변경

Edit files to resolve conflict

Remove id in ProfileTabViewController

shp7724 commented:
잠깐 둘러봤는데 UI가 훨씬 깔끔해졌네요. 다만 "전체 공개 정보 수정" 버튼을 누르면 push가 세번 연속으로 되는 버그가 있습니다.
제가 RxDataSource는 사용을 안해봐서, 회의때 어떻게 구현하신건지 간략하게 설명해주시면 이후에 리뷰해보겠습니다.

shp7724 commented on Facebook/Facebook/Features/New Group/Views/Header&FooterView.swift:
```diff
@@ -0,0 +1,121 @@
+            divider.heightAnchor.constraint(equalToConstant: 0.5),
+            divider.bottomAnchor.constraint(equalTo: self.bottomAnchor),
+            divider.leadingAnchor.constraint(equalTo: self.leadingAnchor, constant: 15),
+            divider.trailingAnchor.constraint(equalTo: self.trailingAnchor, constant: -15),
```
크게 중요한 건 아니지만 여기서 `translatesAutoresizingMaskIntoConstraints`는 밑에서 설정해주고 있어서 지워도 될 것 같아요. (해당 프로퍼티는 클로져 내부에서 설정하는 걸로 통일하는게 좋을 것 같습니다.)

shp7724 commented on Facebook/Facebook/Features/New Group/Views/DetailProfileTableViewCell.swift:
```diff
@@ -7,19 +7,21 @@
     @IBOutlet weak var informationImage: UIImageView!
     @IBOutlet weak var informationLabel: UILabel!
+
+    let disposesBag = DisposeBag()
```
`disposesBag` 오타가 있네요.

shp7724 commented on Facebook/Facebook/Features/New Group/New Group1/EditDetailInformationViewController.swift:
```diff
@@ -0,0 +1,155 @@
+//
```
이 파일들은 `New Group/New Group1` 경로에 있는거라고 나와있는데 xcode 프로젝트상에서는 안보이네요.
필요 없는 파일은 아닌것같은데 어떻게 된 건가요?

shp7724 replied:
@KWSMooBang 요거 확인 부탁드립니다!

shp7724 commented on Facebook/Facebook/Features/New Group/Views/DetailProfileTableViewCell.swift:
```diff
@@ -32,11 +34,38 @@ class DetailProfileTableViewCell: UITableViewCell {
-        }).disposed(by: disposeBag)
+    func configureCell(style: Style) {
+        switch style {
+        case .style1:
+            NSLayoutConstraint.activate([
+                informationImage.heightAnchor.constraint(equalToConstant: 20),
+                informationImage.widthAnchor.constraint(equalToConstant: 20)
+            ])
+        case .style2:
+            NSLayoutConstraint.activate([
+                informationImage.heightAnchor.constraint(equalToConstant: 20),
+                informationImage.widthAnchor.constraint(equalToConstant: 20)
+            ])
```
여기서 style1과 style2로 나누신 이유가 있을까요?
그리고 style1, style2보다는 더 직관적인 이름을 사용하면 좋을 것 같긴 합니다.

KWSMooBang replied:
style1과 style2로 나눈 이유는 동일한 컴포넌트를 가지기만 디자인이 조금씩 다른 cell을 구현하기 위해 나누었습니다. 추후에 더 직관적인 이름으로 바꾸도록 하겠습니다.

shp7724 commented on Facebook/Facebook/Features/New Group/New Group1/SelectInformationViewController.swift:
```diff
@@ -0,0 +1,109 @@
+    
+    private lazy var dataSource = RxTableViewSectionedReloadDataSource<MultipleSectionModel>(configureCell: configureCell)
+    
+    private lazy var configureCell: RxTableViewSectionedReloadDataSource<MultipleSectionModel>.ConfigureCell = { dataSource, tableView, idxPath, _ in
+        switch dataSource[idxPath] {
+        case let .MainProfileItem(profileImage, coverImage, name):
+            let cell = UITableViewCell()
+            
+            return cell
+        case let .ProfileImageItem(image):
+            let cell = UITableViewCell()
+            
+            return cell
+        case let .CoverImageItem(image):
+            let cell = UITableViewCell()
+            
```
비슷한 switch~case 구문이 여러 VC를 걸쳐서 계속 나오고 있는데,
중복을 줄일 수 있는 방법이 없을까요?

yooxdb commented on Facebook/Facebook/Features/New Group/New Group1/EditDetailInformationViewController.swift:
```diff
@@ -0,0 +1,141 @@
+    
+    let disposeBag = DisposeBag()
+    
+    let sections: [MultipleSectionModel] = [
```
<img src="https://user-images.githubusercontent.com/70614553/147870124-097ac3a6-f344-4fde-81bd-b2a81abec67d.png" width="40%" height="40%" align="center">
헤더가 일부 가려지는 버그가 있습니다 확인 부탁드려요!

yooxdb commented on Facebook/Facebook/Features/New Group/New Group1/AddInformationViewController.swift:
```diff
@@ -0,0 +1,218 @@
+    private func isActiveSection() {
+        let sections: [MultipleSectionModel] = [
+            .DetailInformationSection(title: "직장", items: [
+                .SimpleInformationItem(style: .style4, image: UIImage(systemName: "briefcase")!, information: "직장 이름"),
```
페이스북에서 "ㅇㅇㅇㅇㅇ직장 추가"를 누르면 가장 상단의 "직장 이름" textField 칸에는 user가 입력했던 직장 이름이 기본적으로 기재되어 있는데 현재 상태에서는 이 부분이 반영되지 않은 것 같습니다. 확인 부탁드립니다!

yooxdb commented on Facebook/Facebook/Features/New Group/New Group1/SelectInformationViewController.swift:
```diff
@@ -0,0 +1,90 @@
+        }
+    }
+    
+    var inforomationType: String = ""
```
`inforomation`에 오타가 있네요

yooxdb commented on Facebook/Facebook/Features/New Group/New Group1/AddInformationViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        
+        sectionSwitch.rx.tap.bind { [weak self]  in
+            guard let self = self else { return }
+            if self.isActive {
```
switch의 active 상태에 따라 section에는 반영되지만 button의 이미지에는 반영되지 않는 것 같네요 확인 부탁드립니다!

KWSMooBang replied:
이 부분은 버그가 있는 것 같아서 잠시 뒤로 미뤄두었습니다. 나중에 해결하도록 하겠습니다.

yooxdb commented on Facebook/Facebook/Features/New Group/New Group1/AddInformationViewController.swift:
```diff
@@ -0,0 +1,218 @@
+                self.isNotActiveSection()
+                self.isActive = false
+            } else {
+                sectionSwitch.setImage(UIImage(systemName: "checkmark.square.fill")!, for: .normal)
```
systemImage 사용 시에는 forced-unwrapping이 없어도 됩니다 `!` 지우셔도 될거같아요

yooxdb commented on Facebook/Facebook/Features/New Group/New Group1/SelectInformationViewController.swift:
```diff
@@ -0,0 +1,90 @@
+        
+        searchResultBR.bind(to: tableView.rx.items(dataSource: dataSource)).disposed(by: disposeBag)
+        
+        tableView.rx.modelSelected(SectionItem.self).subscribe(onNext: { [weak self] item in
```
사소한 부분이지만 이 부분에 `.disposed` 관련한 내용이 빠진 것 같습니다

yooxdb committed:
Edit Alert Message in LoginVC

- 로그인 오류 시 뜨는 alert 메시지 실제 앱에 맞춰 변경

yooxdb committed:
Edit Alert Message in LoginVC

- 로그인 오류 시 뜨는 alert 메시지 실제 앱에 맞춰 변경

yooxdb committed:
Modify email, pw textfield

- id에서 email로 변수명 변경
- email, pw textfield에 자동 capitalization 비활성화 및 pw textfield 스타일 수정

yooxdb committed:
Modify email, pw textfield

- id에서 email로 변수명 변경
- email, pw textfield에 자동 capitalization 비활성화 및 pw textfield 스타일 수정

yooxdb committed:
Add Login feature

- 로그인 기능 구현
- LoginResponse 모델 수정

yooxdb committed:
Add Login feature

- 로그인 기능 구현
- LoginResponse 모델 수정

yooxdb committed:
Edit SignUpResponse model

- 서버쪽의 수정된 회원가입 response에 맞추어 SignUpResponse 모델 수정

yooxdb committed:
Edit SignUpResponse model

- 서버쪽의 수정된 회원가입 response에 맞추어 SignUpResponse 모델 수정

yooxdb committed:
Refactor signup, login process

- Add AuthManager.swift -> 로그인, 회원가입과 관련한 메소드를 담고 있는 struct(추후 로그아웃도 여기서 구현할 예정)

yooxdb committed:
Refactor signup, login process

- Add AuthManager.swift -> 로그인, 회원가입과 관련한 메소드를 담고 있는 struct(추후 로그아웃도 여기서 구현할 예정)

yooxdb committed:
Chore: 불필요한 코드 정리

yooxdb committed:
Chore: 불필요한 코드 정리

yooxdb committed:
Fix Signup+Login error + update baseURL

yooxdb committed:
Fix Signup+Login error + update baseURL

yooxdb committed:
Resolve conflict

yooxdb committed:
Remove force unwrapping(일부)

yooxdb committed:
Merge remote-tracking branch 'origin/feat/login' into feat/login

Commit:
일부 기능 수정

Commit:
Remove force unwrapping(view controller)

Commit:
함수 길이 조정

yooxdb committed:
Modify LoginResponse

Commit:
ProfileTab CreatePostHeader 추가

yooxdb committed:
AuthResponse로 통합

Commit:
자기 소개 저장 후 dismiss 추가

Commit:
EditProfileView에서 이미지 업로드 후 데이터 재로드

Commit:
생일 변경시 alert 표시 추가

Commit:
다른 사람의 프로필을 볼 경우의 ProfileTab 구현

Commit:
다른 사람의 프로필을 볼 경우의 DetailProfileView 구현

Commit:
Merge remote-tracking branch 'origin/feat/login' into feat/new-profile

Commit:
error fix

Commit:
Merge branch 'feat/profile' into feat/new-profile

Commit:
페이지네이션 모델 로그인 삭제

shp7724 commented:
프로필 이미지를 바꿔도 메인 프로필 이미지가 변하지 않는 것 같습니다(앱 재시작해야 바뀌는 것 확인). 확인 부탁드립니다.

shp7724 commented:
상태 메시지를 편집하면 프로필 사진과 커버 사진이 가로로 늘어집니다.

![image](https://user-images.githubusercontent.com/33917774/148517845-df0df01b-d8ff-46d6-988a-507584366a01.png)

수정 부탁드립니다.

shp7724 commented on Facebook/Facebook/Features/Profile/Controllers/EditProfileViewController.swift:
```diff
@@ -85,104 +241,147 @@ class EditProfileViewController<View: EditProfileView>: UIViewController, UITabl
+                self.presentPicker()
+            }.disposed(by: disposeBag)
+        case 2:
+            sectionButton.rx.tap.bind {
+                if (self.userProfile?.self_intro == nil || self.userProfile?.self_intro == "") {
+                    let addSelfIntroViewController = AddSelfIntroViewController()
+                    let navigationController = UINavigationController(rootViewController: addSelfIntroViewController)
+                    navigationController.modalPresentationStyle = .fullScreen
+                    self.present(navigationController, animated: true, completion: nil)
+                } else {
+                    self.showAlertMenu()
+                }
+            }.disposed(by: disposeBag)
+        case 3:
+            sectionButton.rx.tap.bind {
+                let editDetailInformationViewController = EditDetailInformationViewController()
+                self.push(viewController: editDetailInformationViewController)
+            }.disposed(by: disposeBag)
```
[weak self] 처리가 안되어 있어서 해당 VC가 deinit이 안되고 있습니다. 확인 부탁드려요.

shp7724 reviewed (approved):
수고하셨습니다!

Commit:
프로필 및 커버 이미지 관련 수정(일부), 프로필 탭 기능 수정

Commit:
VC deinit 수정
